### PR TITLE
feat(web-forms#464): prefill default values from url parameter

### LINKED ISSUE #464: Pre-filling using URL parameters
### Description
Support pre-filling form fields using URL parameters with XPath syntax, such as `&d[/data/uuid]=foo`, to match the functionality already available in Enketo. This feature would allow seamless pre-population of form fields (e.g., a text field named uuid) via URL parameters.

Form: 

[pre-fill using url parameter.xlsx](https://github.com/user-attachments/files/21468598/pre-fill.using.url.parameter.xlsx)

---

### Checklist
- **Does it need UI/UX design?**
  - [ ] Yes. Describe the UI/UX requirements or attach a Figma link in the "Design" section.
  - [x] No
- **Does it need API design?**
  - [x] Yes. Specify the endpoints, methods, or data structures needed.
  - [ ] No
- **Does it need design around state flow?**
  - [ ] Yes. Outline the state flow requirements.
  - [x] No
- **Does it need a test plan?**
  - [x] Yes. Add the test plan as a comment in this task or attach the document.
  - [ ] No
---

### User Stories

- **As a** [user type],
  **I want** [functionality],
  **so that** [benefit].

---

### Design

- **Figma**: [Insert URL]
- **Design Docs**: [Insert URL]

---

### Dependencies

-

---

### Additional Notes

- Likely, Central needs to pass these parameters to Web Forms.
- Forum post describing the feature: [How can I specify default values in a web form link?](https://forum.getodk.org/t/how-can-i-specify-default-values-in-a-web-form-link/42847)
- Forum post reporting the missing feature: [Pre-filling using URL parameters does not work in ODK Web Forms](https://forum.getodk.org/t/pre-filling-using-url-parameters-does-not-work-in-odk-web-forms/55939)

---

### Related Issues

-

### PR DESCRIPTION
**Pull request by garethbowen**
Closes getodk/web-forms#464

#### What has been done to verify that this works as intended?

Testing. Test plan to come...

#### Why is this the best possible solution? Were any other approaches considered?

One other option was to create an XML doc and bind that in to the form (a bit like when editing) but because there are multiple ways to set the value it's important to get the order right. This felt clean.

#### How does this change impact users? Describe intentional behavior changes from code updates. What are the regression risks?

There are risks around malicious URLs allowing injection into the form, or modifying protected fields.

#### Does this change require updates to user documentation? If so, please file an issue [here](https://github.com/getodk/docs/issues/new) and include the link below.

Yes. https://github.com/getodk/docs/issues/2080

### CODE REVIEW

File: packages/xforms-engine/src/lib/reactivity/createInstanceValueState.ts [hunk 1]
```diff
@@ -33,9 +34,11 @@ const isEditInitialLoad = (context: ValueContext) => {
 };
 
 const getInitialValue = (context: ValueContext): string => {
-  const sourceNode = context.instanceNode ?? context.definition.template;
-
-  return context.decodeInstanceValue(sourceNode.value);
+  const value =
```
Comment (latin-panda) [inline packages/xforms-engine/src/lib/reactivity/createInstanceValueState.ts:37]: This prefill via URL parameters should not override submissions on edit. If not there already, could you please add a test that covers this scenario?

File: packages/xforms-engine/src/lib/reactivity/createInstanceValueState.ts [hunk 2]
```diff
@@ -33,9 +34,11 @@ const isEditInitialLoad = (context: ValueContext) => {
...
+    getInstanceDefaultValue(context.instanceConfig.instanceDefaults, context) ??
+    context.instanceNode?.value ??
+    context.definition.template?.value;
+  return context.decodeInstanceValue(value);
```
Comment (latin-panda) [inline packages/xforms-engine/src/lib/reactivity/createInstanceValueState.ts:41]: Could you add a test post submission to confirm that the form state is reset and that it prefills using URL parameters?

File: apps/forms/src/components/submission.vue [hunk 3]
```diff
@@ -58,6 +59,7 @@ const enketoId = computed(() => route.params.enketoId ? encodeURIComponent(route
 const st = computed(() => route.query.st ? route.query.st as string : null);
 const useWebForms = computed(() => route.query.webforms === 'true');
 const offline = computed(() => route.params.offline === 'offline');
+const defaultParameters = computed(() => getDefaultParameters(route.query));
```
Comment (latin-panda) [inline apps/forms/src/components/submission.vue]: When in edit submission mode, it shouldn't pass these default parameters.

File: apps/forms/src/components/preview.vue [hunk 4]
```diff
@@ -59,7 +61,7 @@ fetchForm();
       {{ $t('formNotFound') }}
     </div>
     <template v-else-if="webFormsEnabled">
-      <WebFormRenderer :form="form!" :xform="xform!" action-type="preview"/>
+      <WebFormRenderer :form="form!" :xform="xform!" action-type="preview" :default-parameters="defaultParameters" />
```
Comment (latin-panda) [inline apps/forms/src/components/preview.vue:64]: We kinda call this feature "pre-fill from parameters", which could be an alternative name instead of `:default-parameters`

File: apps/forms/test/utils/default-parameters.test.ts [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+      'd[/root/path/that/is/deep]': 'c'
+    };
+    expect(getDefaultParameters(given)).toEqual({
+      '/root/path': 'a',
+      'shortform': 'b',
+      '/root/path/that/is/deep': 'c',
+    });
+  });
+
+  it('returns values with special characters', () => {
```
Comment (latin-panda) [inline apps/forms/test/utils/default-parameters.test.ts:63]: What does the regex match if someone sends something like `d[a][b] `?

File: packages/xforms-engine/src/lib/instance-defaults.ts [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+  return !!PROTECTED_META_FIELDS.find((field) => {
+    return refWithoutRoot === `${OPENROSA_XFORMS_PREFIX}:meta/${OPENROSA_XFORMS_PREFIX}:${field}`;
+  });
+};
+
+export const getInstanceDefaultValue = (
+  defaults: InstanceDefaults,
+  context: ValueContext
+): string | undefined => {
+  if (!defaults || context.contextNode.nodeType === 'attribute') {
```
Comment (latin-panda) [inline packages/xforms-engine/src/lib/instance-defaults.ts]: According to the parameter definition, `defaults` is never undefined or null. So this IF statement can be simplified, or the parameter definition should be `InstanceDefaults | undefined`.

File: packages/xforms-engine/src/lib/instance-defaults.ts [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+    return;
+  }
+  const ref = context.contextReference();
+  const positionOfSecondSlash = ref.indexOf('/', 1);
+  const refWithoutRoot = ref.slice(positionOfSecondSlash + 1);
+  if (isProtectedMetaProperty(refWithoutRoot)) {
+    return;
+  }
+  return defaults[ref] ?? (refWithoutRoot && defaults[refWithoutRoot]) ?? undefined;
```
Comment (latin-panda) [inline packages/xforms-engine/src/lib/instance-defaults.ts:42]: I understand that keys must match the full XPath paths, right?
So the second case is not valid because the root is necessary, isn't it?

File: packages/xforms-engine/src/lib/instance-defaults.ts [hunk 8]
```diff
@@ -0,0 +1,42 @@
+import { OPENROSA_XFORMS_PREFIX } from '@getodk/common/constants/xmlns';
+import type { InstanceValueContext } from '../instance/internal-api/InstanceValueContext';
+import type { AttributeContext } from '../instance/internal-api/AttributeContext';
+import type { InstanceDefaults } from '../client';
+
+type ValueContext = AttributeContext | InstanceValueContext;
+
+const PROTECTED_META_FIELDS = [
```
Comment (latin-panda) [inline packages/xforms-engine/src/lib/instance-defaults.ts:8]: If not already there, could you add tests for these cases?
- Check that the fields with preloaded values are not overridden.  For example, have the deviceID,  pass a URL default for it. And the test asserts the final value is the preload value, not the URL value.
- Now the other way, have instanceName (with no preloaded value), pass a URL default. The test asserts the URL default is rejected.

File: packages/xforms-engine/src/lib/instance-defaults.ts [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+  'deviceID',
+  'deprecatedID',
+  'email',
+  'phoneNumber',
+  'audit',
+];
+
+const isProtectedMetaProperty = (refWithoutRoot: string): boolean => {
+  return !!PROTECTED_META_FIELDS.find((field) => {
+    return refWithoutRoot === `${OPENROSA_XFORMS_PREFIX}:meta/${OPENROSA_XFORMS_PREFIX}:${field}`;
```
Comment (latin-panda) [inline packages/xforms-engine/src/lib/instance-defaults.ts]: We also support meta without a prefix, right?

File: packages/xforms-engine/src/entrypoints/FormInstance.ts [hunk 10]
```diff
@@ -42,6 +42,7 @@ export class FormInstance<
       clientStateFactory: instanceConfig.stateFactory ?? identity,
       computeAttachmentName: instanceConfig.instanceAttachments?.fileNameFactory ?? (() => null),
       preloadProperties: instanceConfig.preloadProperties ?? {},
+      instanceDefaults: instanceConfig.instanceDefaults ?? {},
```
Comment (latin-panda) [inline packages/xforms-engine/src/entrypoints/FormInstance.ts:45]: If it's defined here and coming from InstanceConfig, is it still necessary to have `instanceDefaults` in packages/xforms-engine/src/client/form/LoadForm.ts?

File: packages/xforms-engine/src/client/form/FormInstanceConfig.ts [hunk 11]
```diff
@@ -16,6 +16,8 @@ export interface PreloadProperties {
   readonly phoneNumber?: string;
 }
 
+export interface InstanceDefaults extends Record<string, string> {}
```
Comment (latin-panda) [inline packages/xforms-engine/src/client/form/FormInstanceConfig.ts:25]: The `Record<string, string>` doesn't provide enough information about the expected key shapes. Is there a way to make it clearer? If explaining it with types is hard, a comment will help a lot.

**Review by latin-panda — CHANGES_REQUESTED** [review]
Great work! This feature needs a lot of testing due to many potential scenarios. It's worth investing in unit test coverage to prevent future regressions. 

I haven't completed my review yet, I still need to consider the security, repeats and edge cases. Looking forward to seeing the test plan!

File: apps/forms/src/components/preview.vue (hunk 4 above)
Comment (garethbowen) [reply apps/forms/src/components/preview.vue:64]: I struggled with naming and added a comment to the issue with my rationale for where this ended up: https://github.com/getodk/web-forms/issues/464#issuecomment-5097357558

While we have been calling it "pre-fill", in Enketo it's always "default", and the param is `d` which is short for "default" so I wanted to stay consistent there. I'm definitely open to better naming if you have thoughts!

File: packages/xforms-engine/src/lib/instance-defaults.ts (hunk 7 above)
Comment (garethbowen) [reply packages/xforms-engine/src/lib/instance-defaults.ts:42]: There's a little known and undocumented feature in Enketo where you can leave it off because 99% of the time you're binding into the primary instance so it saves a bit in the URL. So both `d[/root/name]` and `d[name]` are equivalent. I've added this to the related docs PR and there are tests which make sure it works.

File: packages/xforms-engine/src/lib/reactivity/createInstanceValueState.ts (hunk 1 above)
Comment (garethbowen) [reply packages/xforms-engine/src/lib/reactivity/createInstanceValueState.ts:37]: Thanks! Added test, failed, fixed. I decided to just null out the defaults in the engines editInstance function long before it gets here.

File: packages/xforms-engine/src/lib/reactivity/createInstanceValueState.ts (hunk 2 above)
Comment (garethbowen) [reply packages/xforms-engine/src/lib/reactivity/createInstanceValueState.ts:41]: Done.

File: apps/forms/src/components/submission.vue (hunk 3 above)
Comment (garethbowen) [reply apps/forms/src/components/submission.vue]: It's now ignored at the engine level but I've also added a check here so we don't waste the time checking for params that are going to be ignored.

File: apps/forms/test/utils/default-parameters.test.ts (hunk 5 above)
Comment (garethbowen) [reply apps/forms/test/utils/default-parameters.test.ts:63]: Vue parses that to: `{ "d[a][b]": "123" }`
Then the regex matches and sets the default params to: `{ "a][b": "123" }`

Nothing complains but it's never going to match a ref so it's a no-op.

File: packages/xforms-engine/src/lib/instance-defaults.ts (hunk 6 above)
Comment (garethbowen) [reply packages/xforms-engine/src/lib/instance-defaults.ts]: Updated this to check if defaults has any keys so we can short circuit the code if it'll never match (which is the 99% case).

File: packages/xforms-engine/src/lib/instance-defaults.ts (hunk 8 above)
Comment (garethbowen) [reply packages/xforms-engine/src/lib/instance-defaults.ts:8]: I think this is covered by the "instance-defaults" unit test which checks that checking for the default for any of these fields returns undefined, which means the preload works how it always did.

File: packages/xforms-engine/src/lib/instance-defaults.ts (hunk 9 above)
Comment (garethbowen) [reply packages/xforms-engine/src/lib/instance-defaults.ts]: True! I've added a test for the various combinations of namespaces.

File: packages/xforms-engine/src/entrypoints/FormInstance.ts (hunk 10 above)
Comment (garethbowen) [reply packages/xforms-engine/src/entrypoints/FormInstance.ts:45]: No you're right. I think that was left over from a previous idea. Removed now. Thanks!

File: packages/xforms-engine/src/client/form/FormInstanceConfig.ts (hunk 11 above)
Comment (garethbowen) [reply packages/xforms-engine/src/client/form/FormInstanceConfig.ts:25]: Because the key can be either the full xpath or the short form it's difficult to type. I've added a comment. Let me know if that explains it sufficiently!

File: e2e-tests/tests/default-parameters.spec.js [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+  backendClient = new BackendClient(playwright, `${testInfo.project.name}_wf`, projectId);
+  await backendClient.alwaysHideModal();
+  simpleForm = await backendClient.createForm();
+});
+
+test.afterAll(async () => {
+  await backendClient.dispose();
+});
+
+test.describe('Default Parameters', () => {
```
Comment (latin-panda) [inline e2e-tests/tests/default-parameters.spec.js:21]: Could you also assert the preview URL and public links?

File: packages/xforms-engine/test/integration/instance-defaults.test.ts [hunk 13]
```diff
@@ -0,0 +1,183 @@
...
+      scenario.createNewRepeat({ assertCurrentReference: '/root/repeat' });
+      expect(scenario.answerOf('/root/repeat[1]/child')).toEqualAnswer(stringAnswer(''));
+      expect(scenario.answerOf('/root/repeat[2]/child')).toEqualAnswer(stringAnswer('gregory'));
+      expect(scenario.answerOf('/root/repeat[3]/child')).toEqualAnswer(stringAnswer(''));
+    });
+
+    it('ignored instance defaults when editing', async () => {
+      const scenario = await Scenario.init('Bind defaults', formDefinition);
+      scenario.answer('/root/name', 'original');
+      const instanceDefaults = {
```
Comment (latin-panda) [inline packages/xforms-engine/test/integration/instance-defaults.test.ts:115]: Could you please also add a datetime field and another field with an invalid value, such as an invalid geopoint or date, so the codecs work with the prefill feature and help catch future regressions?

**Review by latin-panda — APPROVED** [review]
Excellent! Just a couple of suggestions below. I also checked other scenarios, including repeats, security, memory leaks, and performance.

File: e2e-tests/tests/default-parameters.spec.js (hunk 12 above)
Comment (garethbowen) [reply e2e-tests/tests/default-parameters.spec.js:21]: I don't think the preview URL warrants a full e2e test because I don't think it'll ever be used. The public link is good though, as this is the primary use case. Added.

File: packages/xforms-engine/test/integration/instance-defaults.test.ts (hunk 13 above)
Comment (garethbowen) [reply packages/xforms-engine/test/integration/instance-defaults.test.ts:115]: Good idea. Added.
